packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts

```diff
@@ -1,13 +1,44 @@
+import { CfnBucket, IBucket } from '@aws-cdk/aws-s3';
 import * as cdk from '@aws-cdk/core';
+import { ProductStack } from '../product-stack';
 
 /**
  * Deployment environment for an AWS Service Catalog product stack.
  *
  * Interoperates with the StackSynthesizer of the parent stack.
  */
 export class ProductStackSynthesizer extends cdk.StackSynthesizer {
-  public addFileAsset(_asset: cdk.FileAssetSource): cdk.FileAssetLocation {
-    throw new Error('Service Catalog Product Stacks cannot use Assets');
+  private readonly assetBucket?: IBucket;
+
+  constructor(assetBucket?: IBucket) {
+    super();
+    this.assetBucket = assetBucket;
+  }
+
+  public addFileAsset(asset: cdk.FileAssetSource): cdk.FileAssetLocation {
+    if (!this.assetBucket) {
+      throw new Error('An Asset Bucket must be provided to use Assets');
+    }
+
+    const physicalName = this.physicalNameOfBucket(this.assetBucket);
+
+    (this.boundStack as ProductStack)._addAsset(asset);
```

Comment:
It feels like we should manage the logic for `_addAsset` and `_deployAssets`
here instead of within `ProductStack`.

Reply:
I moved them both over as suggested.